clients/drcachesim/tools/filter/record_filter.cpp

```diff
@@ -302,6 +303,150 @@ record_filter_t::write_trace_entries(per_shard_t *shard,
+        // (without an encoding) in this chunk, unless the user is removing all encodings.
+        // XXX: What if there is a filter removing all encodings but only
+        // to the stop point, so a partial remove that does not change
+        // the filetype?  For now we do not support that, and we re-add
+        // encodings at chunk boundaries regardless.
+        if (TESTANY(OFFLINE_FILE_TYPE_ENCODINGS, per_shard->filetype) &&
+            per_shard->cur_chunk_pcs.find(entry.addr) == per_shard->cur_chunk_pcs.end()) {
+            if (per_shard->pc2encoding.find(entry.addr) == per_shard->pc2encoding.end()) {
+                return "Missing encoding for PC " + std::to_string(entry.addr) +
+                    " at input entry " + std::to_string(per_shard->input_entry_count);
+            }
+            VPRINT(this, 3,
+                   "output new-chunk encoding chunk=%" PRIu64 " ref=%" PRIu64 "\n",
+                   per_shard->chunk_ordinal, per_shard->cur_refs);
+            if (!write_trace_entries(per_shard, per_shard->pc2encoding[entry.addr])) {
+                return "Failed to write";
+            }
+            per_shard->delayed_encodings[entry.addr].clear();
+        }
+    }
+    if (output)
+        per_shard->cur_chunk_pcs.insert(entry.addr);
+    return "";
+}
+
+std::string
+record_filter_t::process_delayed_encodings(per_shard_t *per_shard, trace_entry_t &entry,
+                                           bool output)
+{
+    if (!is_any_instr_type(static_cast<trace_type_t>(entry.type)))
+        return "";
+    if (!output) {
+        if (!per_shard->last_encoding.empty()) {
+            // Overwrite in case the encoding for this pc was already recorded.
+            per_shard->delayed_encodings[entry.addr] =
+                std::move(per_shard->last_encoding);
+        }
+    } else {
+        // Output if we have encodings that haven't yet been output.
+        // We check prev_was_output to rule out filtered-out encodings
+        // (we record all encodings for new-chunk insertion).
+        if (!per_shard->last_encoding.empty() && per_shard->prev_was_output) {
+            // This instruction is accompanied by a preceding encoding. Since
+            // this instruction is not filtered out, output the encoding now.
+            VPRINT(this, 3,
+                   "output just-delayed encoding chunk=%" PRIu64 " ref=%" PRIu64
+                   " pc=0x%zx\n",
+                   per_shard->chunk_ordinal, per_shard->cur_refs, entry.addr);
+            if (!write_trace_entries(per_shard, per_shard->last_encoding)) {
+                return "Failed to write";
+            }
+            // Remove previously delayed encoding that doesn't need to be output
+            // now that we have a more recent version for this instr.
+            per_shard->delayed_encodings.erase(entry.addr);
+        } else if (!per_shard->delayed_encodings[entry.addr].empty()) {
```

Comment:
Re my previous comment here: Similar to the prev_was_output check on L418, do we need to check here whether the delayed encoding was filtered? We probably need a unit test for the scenario where we're filtering out all encodings, the first instance of an instr (which also has encodings prior to it) is filtered out, but its second instance (in the same input and output chunk; with no encodings just prior to it) is not filtered out. Does the current implementation add encodings prior to it?

Reply:
Yes, I added a check for the filetype still having encodings guarding both delayed encoding writes, and add a unit test that does both a trim and removes all encodings.